README.md
# ood_auth_registration

OSC OnDemand Open ID Connect CI Logon Registration page

Summary:

1. display a webform to the user, showing their DN and asking for their OSC credentials
2. upon form submission, bind to ldap to confirm they are who they say they are
3. if successful binding, map the HPC username to the DN using add-user-dn script that apache is given privilege to run on behalf of the user
4. show them success.php, which then redirects the user to their final destination


## Install

See directions here: https://github.com/OSC/Open-OnDemand#authentication-deploy-the-registration-page

## Configuration and Branding

Edit the config.php file to change the branding of the registration and success pages. A global
instance of Config is used in the php that renders these pages.

## Code overview

pages:

* index.php - shows the webform and handles form submission
* whoami.php - for testing purposes, it just shows all of the headers prefixed with `OIDC_CLAIM_`.

resources:

* form.php - web form page template
* success.php - success and redirect page template
* ldap.php - contains function to validate user credentials by binding to ldap
* config.php - branding


CHANGELOG.md
## Unreleased

## 0.4.0 (2019-08-09)

### Changed

- Updated colors to match new OSC branding


## 0.3.0 (2018-10-25)

### Changed

- Updates for new LDAP schema

## 0.2.0 (2017-06-06)

### Added

- Log out link to nav bar
- Display preferred username and issuer in claims displayed to user

### Changed

- For a white list of issuers (OSC keycloak IDPs), we fix the HPC username to map to as being readonly and the value of OIDC_CLAIM_preferred_username

## 0.1.0 (2017-05-26)

### Added

- Add logout link to nav bar

### Changed

- Use new address for OSC ldap server

### Fixed

- Address warnings when fetching claim headers from `$_SERVER` that are not set
- HTML escape claim header values in view
- Correct text references to OpenID Connect

## 0.0.1 (2016-08-05)

### Added:

  - Initial release


form.php
<!DOCTYPE html>
<!--[if lt IE 9]><html class="lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--><html lang="en"><!--<![endif]-->

<?php include('config.php'); ?>

<head>
  <title><?= $config->title ?> | Register OpenID Connect Provider</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">

  <link href="/public/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon" />
  <link rel="stylesheet" href="css/bootstrap.min.css">

<style type="text/css">
.navbar-brand {
  color: <?= $config->nav_brand_text_color ?> !important;
}

.navbar.navbar-inverse {
  background-color: <?= $config->nav_background_color ?> !important;
  border-color: <?= $config->nav_background_color ?> !important;
}

body {
  padding-top: 60px;
}
</style>
</head>

<body>

<!-- navigation bar -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container-fluid">
    <!-- Brand and toggle get grouped for better mobile display -->
    <div class="navbar-header">
      <a class="navbar-brand" href="#"><?= $config->title ?></a>
      <a class="navbar-brand pull-right visible-xs-block" href="/logout">Log Out</a>
    </div>
    <div class="navbar-header pull-right hidden-xs">
      <a class="navbar-brand" href="/logout">Log Out</a>
    </div>
  </div><!-- /.navbar-collapse -->
</nav>
<!-- navbar -->


<!-- content -->
<div class="container">

  <?php if ($error != null): ?>
  <div class="alert alert-danger" role="alert"><strong>Registration failed:</strong> <?= $error ?></div>
  <?php endif; ?>

  <h3>Login to register this OpenID Connect Provider with your HPC account</h3>
  <p>This is the first time you have tried logging into OnDemand with this external
  identity provider. By providing your HPC account credentials we will associate
  the two accounts so that in the future you may login just with this identity provider.</p>

  <div class="row">
    <div class="col-md-6">

   <form class="form-horizontal" action="<?= htmlspecialchars($form_action) ?>" method="post">
     <!-- for IE8 we need to add <fieldset> tags to make styling the legend tag possible
     https://github.com/ssolomon/bootstrap/commit/650ae3b454ecbafb050a3bc7397cf2b03bdb34cd
      -->
     <!--[if lt IE 9]><fieldset><![endif]-->
       <legend>HPC Login</legend>

       <div class="form-group">
         <label class="col-sm-2 control-label" for="username">Username:</label>
         <div class="col-sm-7">
           <input class="form-control" type="text" id="username" name="username"
             <?= $default_user ? 'value="' . $default_user . '" readonly' : 'placeholder="Username"' ?>
           >
         </div>
       </div>

       <div class="form-group">
         <label class="col-sm-2 control-label" for="password">Password:</label>
         <div class="col-sm-7">
           <input class="form-control" type="password" id="password" name="password" placeholder="Password">
         </div>
       </div>

       <div class="form-group">
         <div class="col-sm-offset-2 col-sm-6">
           <button type="submit" class="btn btn-default">Log in</button>
         </div>
       </div>
     <!--[if lt IE 9]></fieldset><![endif]-->
   </form>

    </div>
    <div class="col-md-6">
      <!-- FIXME: address Firefox fieldets issue -->
      <fieldset>
        <legend>OpenID Connect Provider</legend>
          <dl class="dl-horizontal">
            <?php foreach($provider_claims as $key => $value){ ?>
              <dt><?= $key ?></dt><dd><?= htmlentities($value)?></dd>
            <?php } ?>
          </dl>
      </fieldset>
    </div>
  </div>
</div>

<?php if($config->change_password_url){ ?>
<div class="footer">
  <div class="container">
    <hr>
    <p>Forgot password? <a href="<?= $config->change_password_url ?>">Go to <?= $config->change_password_url ?></a> and click "Forgot Password" at the bottom of the page.</p>
  </div>
</div>
<?php } ?>

</body>
</html>


ldap.php
<?php

/**
 * Performs LDAP queries and determines the validity of a user's login
 *
 * Contains the data structures necessary to communicate with LDAP servers.
 * Contains a function that is used by the login page to verify that a certain
 * username/password combination is allowed to log in to OSC systems based on a
 * number of criteria.
 *
 * @category   authentication
 * @package    openid
 * @author     Shaun Brady <shaun.brady@nimbisservices.com>
 * @author     Shawn Hall <shall@osc.edu>
 * @copyright  2014 Ohio Supercomputer Center
 * @version    SVN: $Id:$
 * @link       https://svn.osc.edu/browse/openid
 */

/**
 * LDAP connection settings
 * @name $ldap
 * @global array $GLOBALS['ldap']
 */
$GLOBALS['ldap'] = array (
    # Connection settings
    'primary'         => 'ldaps://ldap1.infra.osc.edu:636',
    'fallback'        => 'ldaps://ldap2.infra.osc.edu:636',
    'protocol'        => 3,
    # AD specific
    'isad'            => false, // are we connecting to Active Directory?
    'lookupcn'        => false, // should we extract CN after the search?
    # Binding account
    'binddn'          => '',
    'password'        => '',
    # User account
    'autodn'          => false, // extract DN from search result, ignore 'testdn'
    'testdn'          => 'cn=%s,ou=People,dc=osc,dc=edu',
    # Searching data
    'searchdn'        => 'ou=People,dc=osc,dc=edu',
    'filter'          => 'uid=%s',

    # Friendly names matching to LDAP attribute names
    'username'        => 'uid',
    'email'           => 'mail',
    'name'            => 'gecos',
    # Note that in LDAP these are loginShell and loginDisabled,
    # but the LDAP library must lowercase these somehow
    'shell'           => 'loginshell',
    'disabled'        => 'logindisabled',
);

/**
 * Custom user info based on LDAP attributes
 * @name $id
 * @global array $GLOBALS['id']
 * DO NOT ENABLE OR EDIT ARRAY ENTRIES! Call to find_ldap() bellow will populate them!
 */
$GLOBALS['id'] = array (
#       'username'              => 'shall',
#       'email'                 => 'shall',
#       'name'                  => 'Shawn Hall',
#       'shell'                 => '/bin/bash',
#       'disabled'              => 'FALSE',
);

/**
 * Search for LDAP account by username. Populate $sreg if found
 * string $username
 */
function find_ldap ($username) {
    global $sreg, $ldap, $profile;

    $no = "no";
    $profile['user_found'] = false;

    if ($username != "") {
        $ds = ldap_connect($ldap['primary']) or $ds = ldap_connect($ldap['fallback']);
        if ($ds) {
            ldap_set_option($ds,LDAP_OPT_PROTOCOL_VERSION,$ldap['protocol']);
            if ($ldap['isad'] == true) ldap_set_option($ds,LDAP_OPT_REFERRALS,0);

            $r = ldap_bind($ds,$ldap['binddn'],$ldap['password']);
            $sr = ldap_search($ds,$ldap['searchdn'],sprintf($ldap['filter'],$username));
            $info = ldap_get_entries($ds, $sr);

            if ($info["count"] == 1) {
                $no = "ok";
                $profile['user_found'] = true;
                if ($ldap['lookupcn'] == true) $profile['auth_cn'] = $info[0]['cn'][0];
                if ($ldap['autodn'] == true) $ldap['testdn'] = $info['0']['dn'];

                # Populate user information from LDAP - if (array_key_exists('keyname', $ldap))...
                $sreg['nickname'] = $info[0][$ldap['nickname']][0];
                $sreg['email']    = $info[0][$ldap['email']][0];

                $values = is_array($ldap['fullname']) ? $ldap['fullname'] : array($ldap['fullname']);
                $fullname = '';
                foreach ($values as $vname) {
                    $aname = $info[0][$vname][0];
                    if ($aname != '') $fullname = ($fullname == '' ? $aname : $fullname . ' ' . $aname);
                }
                $sreg['fullname'] = $fullname;

                $sreg['country']  = $info[0][$ldap['country']][0];

                # Values not obtained from LDAP
                $sreg['language'] = $ldap['def_language'];
                $sreg['postcode'] = $ldap['def_postcode'];
                $sreg['timezone'] = $ldap['def_timezone'];
            }
            ldap_close($ds);
        }
    }
    return $no;
}

/**
 *
 * Tests a username/password combo and returns a message.
 *
 * Perform LDAP bind test with provided username and password
 * Reject user if LDAP attribute that indicates disabled account
 * Reject user if account has expired password
 * Return "ok" if successful auth or an error explanation if not
 *
 * @param string $username  the provided username
 * @param string $password  the provided password
 *
 * @return string  "ok" if user is allowed to log in, or a reason for failure if
 *                 the user is denied
 *
 * @access public
 * @static
 */
function test_ldap ($username, $password) {
    global $ldap;
    // Initially assume user can't log in
    $no = "Invalid username or password.";
    // Disallow empty usernames or passwords
    if (($username != "") && ($password != "")) {
        /* Connect to LDAP servers - this failover probably does NOT work based
        on comments at http://php.net/manual/en/function.ldap-connect.php */
        $ds = ldap_connect($ldap['primary']) or $ds = ldap_connect($ldap['fallback']);
        // If there was a successful connection
        if ($ds) {
            // Set the protocol version
            ldap_set_option($ds,LDAP_OPT_PROTOCOL_VERSION,$ldap['protocol']);
            /* Unused, but good practice to implement. If talking to AD server,
            set necessary options */
            if ($ldap['isad'] == true) ldap_set_option($ds,LDAP_OPT_REFERRALS,0);
            // Attempt to bind by automatically finding DN and using the password
            if ($ldap['autodn'] == true) {
                $r = ldap_bind($ds,$ldap['testdn'],$password);
            }
            else { // otherwise bind explicitly using the username and password
                $r = ldap_bind($ds,sprintf($ldap['testdn'],$username),$password);
            }
            if ($r) { // If it was a successful bind, i.e. a valid username/password combo
                // Query the user in LDAP
                $sr = ldap_search($ds,$ldap['searchdn'],sprintf($ldap['filter'],$username));
                // Parse the returned search results
                $info = ldap_get_entries($ds, $sr);

                // If there is exactly one username match
                if ($info["count"] == 1) {
                    // If automatically finding DN, then set the test DN based on search DN
                    if ($ldap['autodn'] == true) $ldap['testdn'] = $info['0']['dn'];
                    /* Populate user information from LDAP - if (array_key_exists('keyname', $ldap))...
                       Only some of this information is necessary, but thought
                       it would be nice to have if needed in the future */
                       $id['username'] = $info[0][$ldap['username']][0];
                       $id['email']    = $info[0][$ldap['email']][0];
                       $id['name']     = $info[0][$ldap['name']][0];
                       $id['shell']    = $info[0][$ldap['shell']][0];
                    /* Get the LDAP attribute that determines if an account is
                    disabled */
                    $id['disabled'] = $info[0][$ldap['disabled']][0];
                    // If the account is disabled via LDAP attribute or shell, deny access
                    if (strtoupper($id['disabled']) == "TRUE" || $id['shell'] == "/access/denied") {
                        // Let message stay as "Invalid username or password"
                    }
                    /* If the account has an expired password, suggest to the
                       user that they change their password at the appropriate
                       location on the web */
                    elseif ($id['shell'] == "/bin/password_expired") {
                        $no = $id['name'] . ", your password has expired. Please contact OSC Help";
                    }
                    /*
                     * Account locked
                     * 
                     * nsAccountLock is not guaranteed to be present
                     */
                    elseif (array_key_exists('nsaccountlock', $info[0]) && strtoupper($info[0]['nsaccountlock'][0]) == 'TRUE') {
                        $no = $id['name'] . ", your account has been locked. Please contact OSC Help";
                    }
                    // If the user has passed all of these tests, they must be a
                    // valid user that should be allowed access
                    else {
                        $no = "ok";
                    }
                }
            }
            ldap_close($ds); // Close the open LDAP connection
        }
    }
    return $no; // Return "ok" if allowed access, an error message otherwise
}

?>


login.php
<?php
/**
 * Present login page for OSC credentials so we can update grid map file.
 *
 * TODO: add description
 */

require_once('ldap.php');

/**
 * Ensures provided username is valid
 *
 * @param string $username  the provided username
 *
 * @return string  the username if it is valid. An empty string if not valid
 *
 * @access public
 * @static
 */
function scrub_username($username){
    $pattern = '/^[a-zA-Z0-9_\.\-]{1,24}$/';
    if(preg_match($pattern, $username, $matches, PREG_OFFSET_CAPTURE)){
        return $username;
    }
    else{
        return ''; /* short circuits test_ldap() */
    }
}

/**
 * Ensures provided password is valid
 *
 * @param string $password  the provided password
 *
 * @return string  the password
 *
 * @access public
 * @static
 */
function scrub_password($password){
    /* We're more lax on passwords....
       mostly because they can contain almost anything.  This is what
       should be escaped based on:
       http://www.cse.ohio-state.edu/cgi-bin/rfc/rfc2254.html
    */
    //$patterns = array();
    //$patterns[0] = '/\\\/';
    //$patterns[1] = '/\*/';
    //$patterns[2] = '/\(/';
    //$patterns[3] = '/\)/';
    //$patterns[4] = '/\0/';
    //$replacements = array();
    //$replacements[0] = '\\\5c';
    //$replacements[1] = '\\\2a';
    //$replacements[2] = '\\\28';
    //$replacements[3] = '\\\29';
    //$replacements[4] = '\\\00';
    //ksort($patterns);
    //ksort($replacements);
    //return preg_replace($patterns, $replacements, $password);
    return $password;
}

function display_login_form($error = null){
  $redir = "/test";
  $form_action = $_SERVER['SCRIPT_NAME'];
  include "form.php";
}

function add_my_dn($user, $dn, &$error)
{
  $success = true;

  $user = escapeshellarg($user);
  $dn = escapeshellarg($dn);

  $cmd = "/nfs/17/efranz/dev/ood_auth_map/add-user-dn --user {$user} --dn {$dn} 2>&1";
  exec($cmd, $output, $return_var);

  $error = implode("\n", $output);
  if($return_var != 0){
    $success = false;
    $error = implode("\n", $output);
  }

  return $success;
}

// If username_in is set, then the user is attempting to log in
if(isset($_POST['username'])){
    /* Input sanitization mostly done on principle.  I don't believe
       SQL type injection is possible with the library or possibly
       even the protocol.  */
    $error = test_ldap(scrub_username($_POST['username']), scrub_password($_POST['password']));
    // $error = test_ldap("efranz", "xx");
    if ($error == "ok") {
      $add_my_dn_error = null;
      if(! add_my_dn($_POST['username'], $_SERVER['PHP_AUTH_USER'], $add_my_dn_error)){
        echo "<h1>Mapping failed!</h1>";
        echo "<pre>" . $add_my_dn_error . "</pre>";
      }
      else {
        echo "<h1>Mapping successfully created! Redirecting to final destination in 5 seconds...</h1>";
      }
    }
    else {
      display_login_form($error);
    }
}
else {
    display_login_form();
}
?>


manage.php
<?php
/**
 * User self-management page
 * 
 * This page will enable users to self-manage their own mappings.  Currently,
 * the only intended functionality is to view and delete existing mappings.
 * 
 * @author Basil Mohamed Gohar <bgohar@osc.edu> 
 */

require_once __DIR__ . '/config.php';
?>

<!DOCTYPE html>
    <head>
        <title><?= $config->title ?> | Manage OpenID Connect Mappings</title>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">

        <link href="/public/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon" />
        <link rel="stylesheet" href="css/bootstrap.min.css">

        <style type="text/css">
            .navbar-brand {
              color: <?= $config->nav_brand_text_color ?> !important;
            }
            
            .navbar.navbar-inverse {
              background-color: <?= $config->nav_background_color ?> !important;
              border-color: <?= $config->nav_background_color ?> !important;
            }
            
            body {
              padding-top: 60px;
            }
        </style>
    </head>
    <body>
        <!-- navigation bar -->
        <nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
          <div class="container-fluid">
            <!-- Brand and toggle get grouped for better mobile display -->
            <div class="navbar-header">
              <a class="navbar-brand" href="#"><?= $config->title ?></a>
            </div>
            <div class="navbar-header pull-right">
              <a class="navbar-brand" href="/oidc?logout=https%3A%2F%2F<?= htmlentities(urlencode($_SERVER['HTTP_HOST'])) ?>">Logout</a>
            </div>
          </div><!-- /.navbar-collapse -->
        </nav>
        <!-- navbar -->

        <!-- content -->
        <div class="container">

        <h1>Manage User Mappings</h1>
        <p class="lead">View and/or delete your existing mappings</p>
        
          <div class="row">
              <?php
                if (! empty($existing_dns)) {
                    if (isset($_REQUEST['delete']) && isset($_REQUEST['confirm'])) {
                        //  The user is requesting to delete a mapping and has confirmed they are sure..
                        debug($_REQUEST);
                        //  Make a last-ditch validation that the request is valid
                        if (is_valid_dn_mapping($_REQUEST['dn'], $username)) {
                            //  Attempt to delete the mapping
                            debug('We made it this far.');
                            debug(delete_my_dn($username, $_REQUEST['dn'], $error));
                            debug($error);
                            exit;
                        } else {
                            ?>
                            <p class="danger">You have made an invalid request.</p>
                            <?php
                        }
                        exit;
                    } elseif (isset($_REQUEST['delete'])) {
                        //  The user is requesting to delete a mapping but has yet to confirm their decision.
                        debug($_REQUEST);
                        ?>
                        <h2>Delete a mapping <small>Preparing to remove <mark><?=htmlentities($_REQUEST['dn'])?></mark></small></h2>                        
                        <?php
                        print_delete_confirmation_form($_REQUEST['dn'], $username, $password);
                        exit;
                    } else {
                        //  Show them the listing of mappings they have.
                    ?>
                    <div class="col-md-8">
                        <?php print_existing_mappings_form($existing_dns, $_SERVER['PHP_AUTH_USER'], $username, $password); ?>
                    </div>
              <?php
                    }
                }
              ?>
            
          </div>
        </div>
        
        <?php if($config->change_password_url){ ?>
        <div class="footer">
          <div class="container">
            <hr>
            <p>Forgot password? <a href="<?= $config->change_password_url ?>">Go to <?= $config->change_password_url ?></a> and click "Forgot Password" at the bottom of the page.</p>
          </div>
        </div>
        <?php } ?>
    </body>
</html>

success.php
<!DOCTYPE html>
<!--[if lt IE 9]><html class="lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--><html lang="en"><!--<![endif]-->

<?php include('config.php'); ?>

<head>
  <title><?= $config->title ?> | Register OpenID Provider</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="refresh" content="5;url=<?= htmlspecialchars($redir) ?>">

  <link href="/public/favicon.ico" rel="shortcut icon" type="image/vnd.microsoft.icon" />
  <link rel="stylesheet" href="css/bootstrap.min.css">

<style type="text/css">
.navbar-brand {
  color: <?= $config->nav_brand_text_color ?> !important;
}

.navbar.navbar-inverse {
  background-color: <?= $config->nav_background_color ?> !important;
  border-color: <?= $config->nav_background_color ?> !important;
}

body {
  padding-top: 60px;
}
</style>
<body>

<!-- navigation bar -->
<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container-fluid">
    <!-- Brand and toggle get grouped for better mobile display -->
    <div class="navbar-header">
      <a class="navbar-brand" href="/"><?= $config->title ?></a>
    </div>
  </div><!-- /.navbar-collapse -->
</nav>
<!-- navbar -->


<!-- content -->
<div class="container">
  <h3>Successfully registered provider with HPC account</h3>
  <p class="lead">You should be redirected to your destination in 5 seconds. If not, <a href="/">click here to go to the dashboard.</a></p>
</div>

</body>
</html>


index.php
<?php
/**
 * Present login page for OSC credentials so we can update grid map file.
 *
 * TODO: add description
 */

require_once('ldap.php');

/**
 * Fetch a variable from an object without php warnings
 *
 * @param $source The array containing data
 * @param $key The key that will be inspected
 * @param null $default An optional return value if location is empty
 * @return null Return the value of $source[$key] or null if empty
 */
function fetch($source, $key, $default = NULL) {
    return isset($source[$key]) ? $source[$key] : $default;
}

/**
 * Ensures provided username is valid
 *
 * @param string $username  the provided username
 *
 * @return string  the username if it is valid. An empty string if not valid
 *
 * @access public
 * @static
 */
function scrub_username($username){
    $pattern = '/^[a-zA-Z0-9_\.\-]{1,24}$/';
    if(preg_match($pattern, $username, $matches, PREG_OFFSET_CAPTURE)){
        return $username;
    }
    else{
        return ''; /* short circuits test_ldap() */
    }
}

/**
 * Ensures provided password is valid
 *
 * @param string $password  the provided password
 *
 * @return string  the password
 *
 * @access public
 * @static
 */
function scrub_password($password){
    /* We're more lax on passwords....
       mostly because they can contain almost anything.  This is what
       should be escaped based on:
       http://www.cse.ohio-state.edu/cgi-bin/rfc/rfc2254.html
    */
    //$patterns = array();
    //$patterns[0] = '/\\\/';
    //$patterns[1] = '/\*/';
    //$patterns[2] = '/\(/';
    //$patterns[3] = '/\)/';
    //$patterns[4] = '/\0/';
    //$replacements = array();
    //$replacements[0] = '\\\5c';
    //$replacements[1] = '\\\2a';
    //$replacements[2] = '\\\28';
    //$replacements[3] = '\\\29';
    //$replacements[4] = '\\\00';
    //ksort($patterns);
    //ksort($replacements);
    //return preg_replace($patterns, $replacements, $password);
    return $password;
}

// get redirect url, verify it is relative to host
function get_redir(){
  $redir = isset($_GET['redir']) ? $_GET['redir'] : "/";

  // validate
  if($redir && strlen($redir) > 0 && substr($redir, 0, 1) == "/"){
    // do nothing, this is good
  }
  else{
    $redir = "/";
  }

  return $redir;
}

// return preferred username if issuer is in whitelist of issuers that we want
// to use the preferred username for mapping. otherwise return null.
function default_user(){
  $default_user = null;

  $issuer = fetch($_SERVER, "OIDC_CLAIM_iss");
  $issuers_with_default_user = array(
    "https://idp-dev.osc.edu/auth/realms/osc",
    "https://idp-dev.osc.edu/auth/realms/awesim",
    "https://idp-test.osc.edu/auth/realms/osc",
    "https://idp-test.osc.edu/auth/realms/awesim",
    "https://idp.osc.edu/auth/realms/osc",
    "https://idp.osc.edu/auth/realms/awesim",
  );

  if(in_array($issuer, $issuers_with_default_user)){
    $default_user = fetch($_SERVER, "OIDC_CLAIM_preferred_username");
  }

  return $default_user;
}


function display_login_form($error = null){
  $form_action = $_SERVER['REQUEST_URI'];
  $redir = get_redir();
  $default_user = default_user();

  // only display claims that the user would understand
  // default array_filter removes pairs with empty values
  $provider_claims = array_filter(array(
    "Name" => fetch($_SERVER, "OIDC_CLAIM_idp_name"),
    "Issuer" => fetch($_SERVER, "OIDC_CLAIM_iss"),
    "Login ID" => fetch($_SERVER, "OIDC_CLAIM_eppn"),
    "Login User" => fetch($_SERVER, "OIDC_CLAIM_name"),
    "Preferred Username" => fetch($_SERVER, "OIDC_CLAIM_preferred_username"),
    "Login Email" => fetch($_SERVER, "OIDC_CLAIM_email")
  ));

  include "form.php";
}

function display_success_page(){
  $redir = get_redir();
  include "success.php";
}

function add_my_dn($user, $dn, &$error)
{
  $success = true;

  $user = escapeshellarg($user);
  $dn = escapeshellarg($dn);

  $cmd = "/usr/local/bin/add-user-dn --user {$user} --dn {$dn} 2>&1";
  // If the /usr/local/bin/add-user-dn script can't be run by apache, you can copy
  // it locally and run it with this.
  // $cmd = "./add-user-dn --user {$user} --dn {$dn} 2>&1";
  exec($cmd, $output, $return_var);

  $error = implode("\n", $output);
  if($return_var != 0){
    $success = false;
    $error = implode("\n", $output);
  }

  return $success;
}

// If username_in is set, then the user is attempting to log in
if(isset($_POST['username'])){
    /* Input sanitization mostly done on principle.  I don't believe
       SQL type injection is possible with the library or possibly
       even the protocol.  */
    $error = test_ldap(scrub_username($_POST['username']), scrub_password($_POST['password']));
    // $error = test_ldap("efranz", "xx");
    if ($error == "ok") {
      $add_my_dn_error = null;
      if(add_my_dn($_POST['username'], $_SERVER['PHP_AUTH_USER'], $add_my_dn_error)){
        display_success_page();
      }
      else {
        display_login_form("Mapping failed: " . $add_my_dn_error);
      }
    }
    else {
      display_login_form($error);
    }
}
else {
    display_login_form();
}
?>
